Sinemakinci/c make project view docs (#3414)
microsoft/vscode-cmake-tools: docs/cmake-project-view.md

> # CMake Project Outline View

Reviewer: This shouldn't be "CMake Project Status View". But since this is also status bar settings, should this be titled something else, and the file named something else? Like "CMake Options Visibility Configuration" maybe and the file be `cmake-options-config.md`? idk just a suggestion

Starting in CMake Tools 1.16, we will provide users the ability to customize their status bar with related CMake items like presets, kits, variants, build, configure, and launch to allow for users to de-clutter this global status bar as much as they would like to, while also having quick access to commonly performed CMake actions.

Users can specify what is shown in the status bar through their `settings.json` file.

The default settings will be set to the following:
## Default Settings Json
```json
"cmake.status.statusBarVisibility": "hidden"

> "cmake.status.advanced": {

Reviewer: technically there aren't spaces between the settings but it doesn't really matter (nit)

    "build": {
        "statusBarVisibility": "visible"
    },

    "launch": {
        "statusBarVisibility": "visible"
    },

    "debug": {

>         "statusBarVisibility" :"visible"

Reviewer: the space is on the wrong side of the colon

Reviewer: I don't think this one should be resolved yet

    }
}
```

These settings mean that by default all settings will be hidden from the status bar, except for Build, Launch, and Debug actions.

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > StatusBar: Visibility` setting to `Visible`

Reviewer: Also this should be `Cmake > Status: Status Bar Visibility`

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > StatusBar: Visibility` setting to `Visible`

Author: @snehara99 let me know how you think about this update

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > StatusBar: Visibility` setting to `Visible`

Reviewer: I left another comment sorry. since the file updated, I didn't realize these were still there. I'll resolve these. just look at my newest comments

            	"projectStatusVisibility": "visible", "hidden" 
        }, 
        "buildPreset": { 
            	"statusBarVisibility": "visible", "icon", "compact", "hidden" 
            	"statusBarLength": 20 
        }, 
        "buildTarget": { 
            	"statusBarVisibility": "visible", "compact", "hidden" 
            	"statusBarLength": 20 
        }, 
        "ctest": { 
            	"statusBarVisibility": "visible", "icon", "compact", "hidden" 
            	"statusBarLength": 20 
            	"color": true, false 
            	"projectStatusVisibility": "visible", "hidden" 
        }, 
        "testPreset": { 
            	"statusBarVisibility": "visible", "icon", "compact", "hidden" 
            	"statusBarLength": 20
        },
        "launchTarget": { 
            	"statusBarVisibility": "visible", "compact", "hidden" 
            	"statusBarLength": 20 
        }, 
        "debug": { 
            	"statusBarVisibility": "visible", "hidden" 
            	"projectStatusVisibility": "visible", "hidden" 
        },
        "launch": { 
            	"statusBarVisibility": "visible", "hidden" 
            	"projectStatusVisibility": "visible", "hidden" 
        }
}
```

> 

Reviewer: I think a note saying which settings apply for the Project Status View would be helpful, and a clarification on the children of the parent node being affected as well by the visibility setting

Reviewer: This would be folder, configure, build, launch, debug, and ctest

* `folder` is the active folder in your workspace. This is where the project is scoped to.
* `configure` is associated with the CMake Configure action. It can only be found in the Project Status View and can't be seen in the status bar, hence no `statusBarVisibility` property value
* `configurePreset` is associated with the CMake Configure Preset. When selected, the user can modify their active CMake Configure Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)
* `kit` is associated with the active kit selected (when CMake Presets aren't present). When selected, the user can modify their active kit. To learn more about kits, please see [our kit documentation](https://github.com/microsoft/vscode-cmake-tools/docs/kits.md)

> * `variantStatus` is associated with the active variant status (when CMake Presets aren't present). When selected, the user can modify their active variant. To learn more about variants, please see [our variant documentation](https://github.com/microsoft/vscode-cmake-tools/docs/variants.md)

Reviewer: so this is the active variant and the status. they're two things. also maybe worth mentioning that the status doesn't show on the project status view

* `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `CMake: Build Target` from the Command Palette
* `ctest` is associated with running CTest. When selected, it will invoke CTest on the test preset specified. If there is no test preset specified, it will run all your tests by default.
* `testPreset` is associated with selecting your active CMake Test Preset. When selected, the user can modify their active CMake Test Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)
* `launchTarget` is associated with selecting your active launch target. When selected, you can specify the active launch target.
* `debug` is associated with the CMake Debug action. It invokes a debugger on the active launch target.
* `launch` is associated with launching the target. It will run the specified target application in the terminal.

> To reset your CMake options back to the default, hover over your `CMake > Statusbar: Visibility setting` and select `Reset Setting`

Reviewer: This should be something like: 

"To reset your CMake options back to the default, hover over  your `Cmake > Status: Status Bar Visibility` and `Cmake > Status: Advanced` settings, click on the gear icons that appear for more options, and select `Reset Setting` on both.

# CMake Options Visibility Configuration

> Starting in CMake Tools 1.16, we will provide users the ability to customize their status bar with related CMake items like presets, kits, variants, build, configure, and launch to allow for users to de-clutter this global status bar as much as they would like to, while also having quick access to commonly performed CMake actions.

Author: updated, let me know your thoughts!

Users can specify what is shown in the status bar through their `settings.json` file.

The default settings will be set to the following:
## Default Settings Json
```json
"cmake.options.statusBarVisibility": "hidden"
"cmake.options.advanced": {
    "build": {
        "statusBarVisibility": "visible"
    },
    "launch": {
        "statusBarVisibility": "visible"
    },
    "debug": {
        "statusBarVisibility": "visible"
    }
}
```

These settings mean that by default all settings will be hidden from the status bar, except for Build, Launch, and Debug actions.

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: Status Bar: Visibility` setting to `Visible`. This setting is overwritten by`CMake > Options: Advanced`, so if you had added to have some options hidden through this, these options will still remain hidden. 

Reviewer: ```suggestion
To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: Status Bar Visibility` setting to `Visible`. This setting is overwritten by `CMake > Options: Advanced`, so if you had added to have some options hidden through this, these options will still remain hidden. 
```

Reviewer: I learned how to make suggestions, so you don't have to copy paste lol. There was still a tiny typo here.

Reviewer: If you like the suggestions I have, you can directly commit the changes. If you want me to edit my suggestions, I can do that too before you commit so you don't have to do that from the IDE

Reviewer 2: It should be `CMake`, not Cmake

Author: @gcampbell-msft the setting itself says Cmake from some reason. @snehara99 is there any ability to change that?

Reviewer: @gcampbell-msft @sinemakinci1 Yeah there's no way to change that

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: StatusBar: Visibility` setting to `Visible`. This setting is overwritten by`CMake > Options: Advanced setting`, so if you had added to have some options hidden through this, these options will still remain hidden. 

Reviewer: the first setting is `Cmake > Options: Status Bar Visibility` and the second shouldn't have the word "setting" within the ``.
Fixed version:
To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: Status Bar Visibility` setting to `Visible`. This setting is overwritten by `Cmake > Options: Advanced` setting, so if you had added to have some options hidden through this, these options will still remain hidden.

> ![Screenshot of the Visual Studio Code Settings view, with the CMake Status Bar: Visibility options. You can set these to visible, hidden, compact, or icon](images/cmake-statusbar-setting.png)

Reviewer: ```suggestion
![Screenshot of the Visual Studio Code settings view, with the CMake > Status Bar Visibility options. You can set these to visible, hidden, compact, or icon](images/cmake-statusbar-setting.png)
```

Reviewer: shortened to "VS" for consistency with the other two images

> ![Screenshot of the Visaul Studio Code Settings view, with the CMake Statusbar: Visibility options. You can set these to visible, hidden, compact, or icon](images/cmake-statusbar-setting.png)

Reviewer: I think there's some typos here. Fixed:
Screenshot of the Visual Studio Code settings view, with the Cmake > Options: Status Bar Visibility options. You can set these to visible, hidden, compact, or icon.

## Configuring your CMake Status Bar

>  You can configure settings for each of the following CMake actions in your settings.json to either be `visible` or `hidden` in the status bar through the `statusBarVisibility` variable.  Settings that are available to be hidden from the project status view side bar can be configured to be `visible` or `hidden` through the `projectStatusVisibility` variable.

Reviewer 2: ```suggestion
You 
```

>  You can configure settings for each of the following CMake actions in your settings.json to either be `visible` or `hidden` in the status bar through the `statusBarVisibility` variable.  Settings that are available to be hidden from the project status view side bar can be configured to be `visible` or `hidden` through the `projectStatusVisibility` variable.

Reviewer 2: ```suggestion
You can configure settings for each of the following CMake actions in your settings.json to either be `visible` or `hidden` in the status bar through the `statusBarVisibility` variable.  Settings that are available to be hidden from the project status view side bar can be configured to be `visible` or `hidden` through the `projectStatusVisibility` variable.
```

Nit: Remove space at the beginning.

 
To make options visible in the status bar take up less space, you can configure certain options to be only the respective icon through the `icon` option in `statusBarVisibility` or specify a given character length through the `compact` option in `statusBarVisibility`. If a `statusBarVisibility` option is set to `compact` you can then specify an integer length for how many characters you want an option to take up through the `statusBarLength` option. It will truncate your existing status bar option to that specified character length. The default option is 20 characters.

You can also configure options to be `visible` or `hidden` in the Project Status View in the CMake Tools sidebar. The options that allow for this customization are:
*`folder`
*`configure`
*`build`
*`ctest`
*`debug`
*`launch`

> Note: if you set one of these to hidden, the parent node will entirely be hidden. For example, if you set `build` to `hidden` in `projectStatusVisibility`, the option to select your build preset will also be hidden

Reviewer: ```suggestion
Note: if you set one of these to hidden, the parent node and its child nodes will be hidden. For example, if you set `build` to `hidden` in `projectStatusVisibility`, the option to select your build preset will also be hidden
```

Reviewer: also just a suggestion; feel free to ignore

Reviewer 2: I would suggest `and it's children`.

 
The full level of options for the CMake status can be seen below:

> ### CMake Status Bar Configuration Options in your Settings Json

Reviewer: ```suggestion
### CMake Status Bar & Project Status View Configuration Options in your Settings Json
```

Reviewer: you can ignore this one. just an idea

Reviewer 2: ```suggestion
### CMake Status Bar and Project Status View Configuration Options in your Settings Json
```

I like this suggestion but I think (I could be wrong) that the common practice is to spell out the word "and".

```json

> "cmake.options.statusBarVisibility": "visibile", "icon", "compact", "hidden" 

Reviewer: visible is spelled wrong here "visible"

            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "buildTarget": { 
            "statusBarVisibility": "visible", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "ctest": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20, 
            "color": true, false, 
            "projectStatusVisibility": "visible", "hidden" 
        }, 
        "testPreset": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20
        },
        "launchTarget": { 
            "statusBarVisibility": "visible", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "debug": { 
            "statusBarVisibility": "visible", "hidden", 
            "projectStatusVisibility": "visible", "hidden" 
        },
        "launch": {
            "statusBarVisibility": "visible", "hidden", 
            "projectStatusVisibility": "visible", "hidden" 
        }
}
```

* `folder` is the active folder in your workspace. This is where the project is scoped to.

> * `configure` is associated with the CMake Configure action. It can only be found in the Project Status View and can't be seen in the status bar, hence no `statusBarVisibility` property value

Reviewer: needs a period at the end (nit)

> * `configurePreset` is associated with the CMake Configure Preset. When selected, the user can modify their active CMake Configure Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)

Reviewer: needs a period at the end (nit)

> * `kit` is associated with the active kit selected (when CMake Presets aren't present). When selected, the user can modify their active kit. To learn more about kits, please see [our kit documentation](https://github.com/microsoft/vscode-cmake-tools/docs/kits.md)

Reviewer: needs a period at the end (nit)

> * `variant` is associated with the active variant (when CMake Presets aren't present). When selected, the user can modify their active variant. Variant status does not show on the Project Status View, but will show in the status bar when set to `visible`. To learn more about variants, please see [our variant documentation](https://github.com/microsoft/vscode-cmake-tools/docs/variants.md)

Reviewer: needs a period at the end (nit)

* `build` is associated with the CMake Build action. It invokes a CMake build on your build target using your build preset or variant.

> * `buildPreset` is associated with your active CMake Build Preset. When selected, the user can modify their active CMake Build Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)

Reviewer: needs a period at the end (nit)

> * `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `CMake: Build Target` from the Command Palette.

Reviewer: ```suggestion
* `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `Cmake: Build Target` from the Command Palette.
```

Reviewer 2: This should be `CMake`, with a capital M. So I don't think this suggestion should be taken.

Author: yeah, its still a capital M in the command palette, just not in the settings

> * `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `CMake: Build Target` from the Command Palette

Reviewer: needs a period at the end (nit)

* `ctest` is associated with running CTest. When selected, it will invoke CTest on the test preset specified. If there is no test preset specified, it will run all your tests by default.

> * `testPreset` is associated with selecting your active CMake Test Preset. When selected, the user can modify their active CMake Test Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)

Reviewer: needs a period at the end (nit)

* `launchTarget` is associated with selecting your active launch target. When selected, you can specify the active launch target.
* `debug` is associated with the CMake Debug action. It invokes a debugger on the active launch target.
* `launch` is associated with launching the target. It will run the specified target application in the terminal.

> To reset your CMake options back to the default, hover over your `CMake > Options: Status bar Visibility` and `CMake > Options: Advanced` settings and select the gear icons that appear for more options. From there, select `Reset Setting` on both options

Reviewer: bar should be capitalized and the m in both CMakes need to be lowercase like Cmake. Also needs a period at the end.
Fixed:
To reset your CMake options back to the default, hover over your `Cmake > Options: Status Bar Visibility` and `Cmake > Options: Advanced` settings and select the gear icons that appear for more options. From there, select `Reset Setting` on both options.

> ![Screenshot of the VS Code Setting option to Reset Setting to the left of CMake Statusbar visibility](images/cmake-setting-2.png)

Reviewer: The setting should be Cmake > Options: Status Bar Visibility

> ![Screenshot of the VS Code Setting option to Reset Setting to the left of Cmake Statusbar visibility](images/cmake-setting-2.png)

Reviewer: ```suggestion
![Screenshot of the VS Code setting option to Reset Setting to the left of Cmake > Options: Status Bar Visibility](images/cmake-setting-2.png)
```

Reviewer 2: It should be `CMake` not Cmake

> ![Screenshot of the VS Code Setting option to Reset Setting to the left of Cmake Statusbar visibility](images/cmake-setting-2.png)

Reviewer 2: ```suggestion
![Screenshot of the VS Code Setting option to Reset Setting to the left of CMake Statusbar visibility](images/cmake-setting-2.png)
```

All references of CMake should be spelled with capital CM

> ![Screenshot of the VS Code Setting option to Reset Setting to the left of CMake Options Advanced visibility](images/cmake-setting-1.png)

Reviewer: The setting should be Cmake > Options: Advanced

> ![Screenshot of the VS Code Setting option to Reset Setting to the left of Cmake Options Advanced visibility](images/cmake-setting-1.png)

Reviewer: ```suggestion
![Screenshot of the VS Code setting option to Reset Setting to the left of Cmake > Options: Advanced visibility](images/cmake-setting-1.png)
```

Reviewer 2: Same comment here, it should be `CMake`

# CMake Options Visibility Configuration

Starting in CMake Tools 1.16, we will provide users the ability to customize their status bar with related CMake items like presets, kits, variants, build, configure, and launch to allow for users to de-clutter this global status bar as much as they would like to, while also having quick access to commonly performed CMake actions.

Users can specify what is shown in the status bar through their `settings.json` file.

The default settings will be set to the following:
## Default Settings Json
```json
"cmake.options.statusBarVisibility": "hidden"
"cmake.options.advanced": {
    "build": {
        "statusBarVisibility": "visible"
    },

> 

Reviewer: there usually arent spaces here either, but also doesn't really matter (sorry for nit)

    "launch": {
        "statusBarVisibility": "visible"
    },

    "debug": {
        "statusBarVisibility": "visible"
    }
}
```

These settings mean that by default all settings will be hidden from the status bar, except for Build, Launch, and Debug actions.

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Status: StatusBar: Visibility` setting to `Visible`. This setting is overwritten by your CMake status settings in your `settings.json`, so if you had added to have some options hidden through your `settings.json`, these options will still remain hidden. 

Reviewer: This should now be `Cmake > Options: Status Bar Visibility`

Reviewer: this setting is specifically overwritten by `Cmake > Options: Advanced` settings. doesn't matter if it's settings.json or not. That's just where they happen to edit it.

You can also configure options to be `visible` or `hidden` in the Project Status View in the CMake Tools sidebar. The options that allow for this customization are:
*`folder`
*`configure`
*`build`
*`ctest`
*`debug`
*`launch`
Note: if you set one of these to hidden, the parent node will entirely be hidden. For example, if you set `build` to `hidden` in `projectStatusVisibility`, the option to select your build preset will also be hidden
 
The full level of options for the CMake status can be seen below:

### CMake Status Bar Configuration Options in your Settings Json
```json
"cmake.options.statusBarVisibility": "visibile", "icon", "compact", "hidden" 
"cmake.options.advanced": { 

        "folder": { 

>         	"statusBarVisibility": "visible", "icon", "compact", "hidden", 

Reviewer: this tab is off (nit)

            "statusBarLength": 20 
        }, 
        "variant": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden" 
        }, 
        "build": { 
            "statusBarVisibility": "visible", "icon", "hidden",
            "projectStatusVisibility": "visible", "hidden" 
        }, 
        "buildPreset": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "buildTarget": { 
            "statusBarVisibility": "visible", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "ctest": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20, 
            "color": true, false, 
            "projectStatusVisibility": "visible", "hidden" 
        }, 
        "testPreset": { 
            "statusBarVisibility": "visible", "icon", "compact", "hidden", 
            "statusBarLength": 20
        },
        "launchTarget": { 
            "statusBarVisibility": "visible", "compact", "hidden", 
            "statusBarLength": 20 
        }, 
        "debug": { 

>         	"statusBarVisibility": "visible", "hidden", 

Reviewer: the tabs for debug and launch options are off too (nit)

* `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `CMake: Build Target` from the Command Palette
* `ctest` is associated with running CTest. When selected, it will invoke CTest on the test preset specified. If there is no test preset specified, it will run all your tests by default.
* `testPreset` is associated with selecting your active CMake Test Preset. When selected, the user can modify their active CMake Test Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)
* `launchTarget` is associated with selecting your active launch target. When selected, you can specify the active launch target.
* `debug` is associated with the CMake Debug action. It invokes a debugger on the active launch target.
* `launch` is associated with launching the target. It will run the specified target application in the terminal.

> To reset your CMake options back to the default, hover over your `CMake > Status: Status bar Visibility` and `CMake: Status: Advanced` settings and select the gear icons that appear for more options. From there, select `Reset Setting` on both options

Reviewer: Here too: `Cmake > Options: Status Bar Visibility` and `Cmake > Options: Advanced`

# CMake Options Visibility Configuration

Starting in CMake Tools 1.16, we will provide users the ability to customize their status bar with related CMake items like presets, kits, variants, build, configure, and launch to allow for users to de-clutter this global status bar as much as they would like to, while also having quick access to commonly performed CMake actions. By default, only `Build`, `Debug`, and `Run` will persist in the status bar. All of the CMake actions for configuring your project will now be available in the CMake Tools side bar under **Project Status** View.

![Screenshot of the CMake Tools side bar to the left in VS Code with the Project Status View open](images/cmake-tools-sidebar.png)

Users can specify what is shown in the status bar through their `settings.json` file.

The default settings will be set to the following:

> ## Default Settings Json

Reviewer: ```suggestion

## Default Settings Json
```

Reviewer: I just added a blank line before the title

```json
"cmake.options.statusBarVisibility": "hidden"
"cmake.options.advanced": {
    "build": {
        "statusBarVisibility": "visible"
    },
    "launch": {
        "statusBarVisibility": "visible"
    },
    "debug": {
        "statusBarVisibility": "visible"
    }
}
```

These settings mean that by default all settings will be hidden from the status bar, except for Build, Launch, and Debug actions.

> To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: Status Bar Visibility` setting to `Visible`. This setting is overwritten by `CMake > Options: Advanced`, so if you had added to have some options hidden through this, these options will still remain hidden. 

Reviewer: ```suggestion
To revert to the prior experience and have all of your presets display in the status bar, set your `Cmake > Options: Status Bar Visibility` setting to `visible`. This setting is overwritten by `Cmake > Options: Advanced`, so if you had added to have some options hidden through this, these options will still remain hidden. 
```

![Screenshot of the VS Code Settings view, with the CMake > Status Bar Visibility options. You can set these to visible, hidden, compact, or icon](images/cmake-statusbar-setting.png)

> ## Configuring your CMake Status Bar

Reviewer: ```suggestion
## Configuring your CMake Status Bar and Project Status View
```

Reviewer: for consistency; i missed this last time sorry

* `build` is associated with the CMake Build action. It invokes a CMake build on your build target using your build preset or variant.
* `buildPreset` is associated with your active CMake Build Preset. When selected, the user can modify their active CMake Build Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project).
* `buildTarget` is associated with your active CMake Build Target. When selected, the user can specify their active Build Target. This will be the target invoked when the user presses the `Build` icon in the status bar (if not hidden) or runs `CMake: Build Target` from the Command Palette.
* `ctest` is associated with running CTest. When selected, it will invoke CTest on the test preset specified. If there is no test preset specified, it will run all your tests by default.

> * `testPreset` is associated with selecting your active CMake Test Preset. When selected, the user can modify their active CMake Test Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project)

Reviewer: ```suggestion
* `testPreset` is associated with selecting your active CMake Test Preset. When selected, the user can modify their active CMake Test Preset from the list detected in their CMakePresets.json and CMakeUserPresets.json (if found in project).
```

Reviewer: (nit)
